fn/mvar.go

```diff
@@ -0,0 +1,212 @@
+package fn
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// MVar[A any] is a structure that is designed to store a single value in an API
+// that dispenses with data races. Think of it as a box for a value.
+//
+// It has two states: full and empty.
+//
+// It supports two operations: take and put.
+//
+// The state transition rules are as follows:
+//  1. put while full blocks.
+//  2. put while empty sets.
+//  3. take while full resets.
+//  4. take while empty blocks.
+//  5. read while full nops.
+//  6. read while empty blocks.
+type MVar[A any] struct {
```

Comment:
Current revision of this PR doesn't make use of this API. I'm leaving it here until this PR is finalized and if it doesn't make the final cut, I can drop the commit entirely or I can PR separately if we think it is useful.